### PR TITLE
Upload UI tests

### LINKED ISSUE #705: Implement upload tests
As suggested by @tobias47n9e after the 2.4 upload crash fiasco. We probably need lots of other tests too, but these ones should be a higher priority IMO.

### PR DESCRIPTION
**Pull request by domdomegg**
**Description**

Fixes #705 (as I believe #2086 did unit tests, and this does UI tests so it should be all covered)

- Increasing number of tests should decrease number of bugs
- Key component of app is upload, this Espresso-based test validates it

Would hopefully catch issues like #2629

**TODO**

Having the test work automatically from a fresh install. Currently it requires running the app and logging in, and having a file called `image.jpg` in the base of the filesystem.

I'm not sure how to go about this, (hence the 'help needed'). I think this roughly breaks into two fairly separate parts:
- Automatically logging in with credentials securely stored in Travis (not publicly on Github). Maybe this can be done in a similar way to how the APK signing keys work?
- Updating the android image to have a sample `image.jpg` in `/storage/emulated/0/` (symlink'ed from `/mnt/sdcard` and `/sdcard/`)

Also I'm worried that if the API is very slow, the delays I've put in the test at certain points to wait for responses won't be long enough - I'm not sure what we can do about this too.

These will need to be done to get this working on Travis, however as it's not actually running the Android tests at the moment we could merge anyways and fix this when we re-enable them.

**Contributing**

If you think you can help please do create a branch continuing mine and submit PRs back to `domdomegg/test-upload` :)

### CODE REVIEW

**Comment by maskaravivek** [comment]
Great initiative @domdomegg! I can try helping with some of the things that you have mentioned in the help section. 

> logging in

It should be quite simple. We can probably get a test user created as discussed in #2226. 

> and having a file called image.jpg

Yes, this should be possible via travis. 

Can you resolve merge conflicts so that i can work on top of it?

**Comment by domdomegg** [comment]
The idea in #2226 was to create a locked down user which we could share the password with so it could run instrumentation tests automatically on anyone's machine - for example tests on the contributions activity. I should have updated it before, but it doesn't seem possible as you can't block a user from changing their password.  Also if a user is locked down it means we can't upload the test image.

Therefore I think we need a new test user which isn't locked down, but we just keep the login details secretly stored on Travis, and then expect people running tests requiring login on their own machines to login to their own test accounts.

**Comment by vanshikaarora** [comment]
Hey @domdomegg , I want to contribute can you please tell about few test that you are not working upon so that I can write the unit tests :)

**Comment by domdomegg** [comment]
> Hey @domdomegg , I want to contribute can you please tell about few test that you are not working upon so that I can write the unit tests :)

I think this PR requires quite a few Travis config changes which only a few people have access to. I've just submitted #2636 which is a WIP on WelcomeActivityTest (the tutorial) with some TODO ideas there I'm not currently working on.

**Comment by vanshikaarora** [comment]
@domdomegg As far I have noticed the previous tests were written in Kotlin, so from now onwards are we writing the unit tests in java or in kotlin?

**Comment by domdomegg** [comment]
> @domdomegg As far I have noticed the previous tests were written in Kotlin, so from now onwards are we writing the unit tests in java or in kotlin?

Good point actually. All the unit tests are done in Kotlin, and the one instrumentation (android) test was done in Java so I continued in Java. Actually thinking about it properly I think we should use Kotlin for the instrumentation tests so all the tests are the same language.

**Comment by maskaravivek** [comment]
> Updating the android image to have a sample image.jpg in /storage/emulated/0/ (symlink'ed from /mnt/sdcard and /sdcard/)

@domdomegg I have raised a PR to fix this. Now in the setup phase, I generate an image programmatically and then use it for uploading. 

https://github.com/domdomegg/apps-android-commons/pull/1

**Comment by maskaravivek** [comment]
Another thing that we need to take care of is: 

Runtime Permissions. As of now, I had to grant permission beforehand.

File: app/src/androidTest/java/fr/free/nrw/commons/UploadTest.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        onView(withId(R.id.submit))
+                .perform(click());
+
+        try {
+            Thread.sleep(10000);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+        String fileUrl = "https://commons.wikimedia.beta.wmflabs.org/wiki/File:" +
```
Comment (maskaravivek) [inline app/src/androidTest/java/fr/free/nrw/commons/UploadTest.java]: Can we enhance this bit as well? We can probably call media wiki api to assert if the image was uploaded.

**Comment by maskaravivek** [comment]
Apart from this, I am working on automating the login bit. Do we have other action items for making this master ready?

**Comment by domdomegg** [comment]
> Apart from this, I am working on automating the login bit. Do we have other action items for making this master ready?

I don't *think* so. There probably will be more annoying issues once we enable it in Travis though.

**Comment by maskaravivek** [comment]
> I don't think so. There probably will be more annoying issues once we enable it in Travis though.

Yes, the final goal should be to have everything work on a fresh install itself.

**Comment by misaochan** [comment]
Wow, thanks @domdomegg ! :) 

As the tests are run on Commons beta, I personally think it is OK to use a regular beta account. This:

> Therefore I think we need a new test user which isn't locked down, but we just keep the login details secretly stored on Travis, and then expect people running tests requiring login on their own machines to login to their own test accounts.

Sounds like a good idea to me.

**Comment by maskaravivek** [comment]
@domdomegg Have raised https://github.com/domdomegg/apps-android-commons/pull/2 to automate the login process.

**Comment by maskaravivek** [comment]
@domdomegg Can we merge this PR now?

**Comment by domdomegg** [comment]
> @domdomegg Can we merge this PR now?

I think so

**Pull request merged**